### PR TITLE
feat(multiple_testing_power): make input_utils support multiple metrics

### PR DESCRIPTION
**Pull request by xyzou685**
To help us get this pull request reviewed and merged quickly, please be sure to include the following items:

* [x] Tests (if applicable)
* [ ] Documentation (if applicable)
* [ ] A full explanation here in the PR description of the work done and statistical methods used

## PR Type
What kind of change does this PR introduce?

* [ ] Bugfix
* [x] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] Tests
* [ ] Other

## Backward Compatibility

Is this change backward compatible with the most recently released version? Does it introduce changes which might change the user experience in any way? Does it alter the API in any way?

* [ ] Yes (backward compatible)
* [ ] No (breaking changes)

### CODE REVIEW

**Comment by xyzou685** [comment]
Made changes to `sample_size_calculator` and its test just for the consistency purpose. Please note that changes like replacing `register_bool_metric`, `register_numeric_metric`, `register_ratio_metric` by `register_metrics` are out of the scope of this PR

File: sample_size/sample_size_calculator.py
Comment (pchiang-godaddy) [inline sample_size/sample_size_calculator.py:94]: confused by this - metrics number !== number of variants right?

Comment (xyzou685) [reply sample_size/sample_size_calculator.py:94]: Oh that's right, it should be `for i in range(len(metric_type))` instead. `variants` should be used only when `MultipleTesting` simulates p-values

File: sample_size/scripts/sample_size_run.py
Comment (dfu-godaddy) [inline sample_size/scripts/sample_size_run.py:26]:
```suggestion
        metrics = get_metrics()
        calculator.register_metric(metrics)
```

File: sample_size/scripts/input_utils.py
Comment (dfu-godaddy) [inline sample_size/scripts/input_utils.py:106]: We can get one list `metrics` with all data we need to register metrics. 
```suggestion
def get_metrics() -> List[Dict[str, Any]]:
```

Comment (pchiang-godaddy) [reply sample_size/scripts/input_utils.py:106]: +1 on this, just make one input parameter with a list of two keys 
```python
[{"metric_type": "", "metric_metadata": <metadata>}]
```

File: sample_size/scripts/input_utils.py
Comment (pchiang-godaddy) [inline sample_size/scripts/input_utils.py:83]:
```suggestion
        input("Enter the number of cohorts for this test. \n" "definition: Control + number of treatments: ").strip().lower()
```

File: sample_size/scripts/input_utils.py
Comment (pchiang-godaddy) [inline sample_size/scripts/input_utils.py:89]: can you also add `Press Enter to use default variant = 2 if you have only 1 control and 1 treatment`

File: sample_size/scripts/sample_size_run.py
Comment (dfu-godaddy) [inline sample_size/scripts/sample_size_run.py:24]: the second attribute should be power. You will need to update SampleSizeCalculator to take variants.

File: sample_size/scripts/input_utils.py
Comment (vjohnson1-godaddy) [inline sample_size/scripts/input_utils.py:6]: Is there a reason we're not using click?  Seems like a great use-case

File: tests/sample_size/scripts/test_input_utils.py
Comment (xyzou685) [reply tests/sample_size/scripts/test_input_utils.py:260]: what's the difference between the two expression? I used to think `assert register` would be more strict as in this case `register` has to be a boolean and its value is `True`

**Comment by pchiang-godaddy** [comment]
Reminder: when you merge all the PRs and the package is ready to release, please follow the RELEASE guide here to make a release PR, and I can walk you through the `Create a new release from Github Page` step after that release PR is merged https://github.com/godaddy/sample-size/blob/main/RELEASE.md#issuing-a-new-release

File: tests/sample_size/scripts/test_input_utils.py
Comment (pchiang-godaddy) [reply tests/sample_size/scripts/test_input_utils.py]: It is best practice to use the same pattern across the file I think. Plus using assert won't print out clear output. see https://stackoverflow.com/questions/33491975/why-not-use-pythons-assert-statement-in-tests-these-days

**Comment by xyzou685** [comment]
> Reminder: when you merge all the PRs and the package is ready to release, please follow the RELEASE guide here to make a release PR, and I can walk you through the `Create a new release from Github Page` step after that release PR is merged https://github.com/godaddy/sample-size/blob/main/RELEASE.md#issuing-a-new-release

Thanks for reminding, I'll create this PR at the end

File: sample_size/sample_size_calculator.py
Comment (pchiang-godaddy) [inline sample_size/sample_size_calculator.py:58]: you probably miss the suggestion that Dannie commented, would be good. to validate the. metadata:
image

**Comment by pchiang-godaddy** [comment]
this you merge other PRs, probably makes more sense to compare with `main` now?
image

File: sample_size/scripts/sample_size_run.py
Comment (vjohnson1-godaddy) [inline sample_size/scripts/sample_size_run.py:34]: nit: FWIW, you should avoid excepting with `Exception`, but it's not a blocker

File: tests/sample_size/test_sample_size_calculator.py
Comment (vjohnson1-godaddy) [inline tests/sample_size/test_sample_size_calculator.py:85]: Can you add a comment (and a JIRA ticket if you feel like it) to eventually parameterize these `get_overall` and  `register` test cases?

File: sample_size/scripts/input_utils.py
Comment (vjohnson1-godaddy) [inline sample_size/scripts/input_utils.py:121]:
```suggestion
def get_metrics() -> List[Dict[str, Any]]:
    metrics = []
    while register_another_metric():
        metric_type = get_metric_type()
        metric_metadata = get_metric_parameters(METRIC_PARAMETERS[metric_type])
        metric_metadata["mde"] = get_mde(metric_type)
        metrics.append({"metric_type": metric_type, "metric_metadata": metric_metadata})
```

**Comment by xyzou685** [comment]
> Looks like the previous PR is merged to a specific branch, are you going to create another PR to merge to main branch?

Yes I am trying to merge everything to `multiple-testing-calculator`, then have it merged to `main`

**Review by vjohnson1-godaddy — COMMENTED** [review]
I'll leave final approvals to the other reviewers after you add the schema validation, but looks great! Could definitely leverage paramterization in the test class, but we can cross that bridge another day.

File: sample_size/scripts/input_utils.py
Comment (xyzou685) [reply sample_size/scripts/input_utils.py]: Isn't this going to ask users 'Are you going to register another metric? (y/n)' before they enter the first metric?

File: sample_size/sample_size_calculator.py
Comment (xyzou685) [reply sample_size/sample_size_calculator.py:62]: just did. Here we are validating metrics, a list of dictionaries, instead of the whole experiment config as in the reference, correct?

Comment (dfu-godaddy) [reply sample_size/sample_size_calculator.py:62]: correct. You can move the `METRICS_SCHEMA` outside of `validate_schema`. It is a constant and can be put together with the default parameters at the top of this file.

**Review by dfu-godaddy — APPROVED** [review]
looks good to me with some nits.

File: sample_size/scripts/input_utils.py
Comment (dpandey-godaddy) [reply sample_size/scripts/input_utils.py:115]: > # TODO: add the case of registering the first metric
This can be handled by emulating a do while loop

File: sample_size/scripts/sample_size_run.py
Comment (dpandey-godaddy) [inline sample_size/scripts/sample_size_run.py:26]: Do we need to update Readme file now that we will support multiple metrics?

Comment (xyzou685) [reply sample_size/scripts/sample_size_run.py:26]: Yes I'm planning on doing that in my last PR(there's one more PR coming for the code change)

Comment (dpandey-godaddy) [reply sample_size/scripts/sample_size_run.py:26]: sounds good

File: sample_size/scripts/input_utils.py
Comment (dpandey-godaddy) [reply sample_size/scripts/input_utils.py:6]: yeah, let's at least have a TODO if we are planning on doing it later

File: sample_size/sample_size_calculator.py
Comment (dpandey-godaddy) [inline sample_size/sample_size_calculator.py:16]: should we put this in a `schemas/metrics_schema.json` and load it instead of putting here?

File: sample_size/scripts/input_utils.py
Comment (pchiang-godaddy) [reply sample_size/scripts/input_utils.py:115]: @xyzou685 are you planning to fix this by another PR? I would suggest either making change in this PR or just create another one only for that change. Instead of putting it with the `merge multiple-testing-calculator to main` PR you will make at the end otherwise it will be hard to know the exact new change that is not approved yet.

File: sample_size/sample_size_calculator.py
Comment (xyzou685) [reply sample_size/sample_size_calculator.py]: Do we have to make a schema folder though there's only one schema in total? Can I put the file with other script like the latest commit does?

File: sample_size/scripts/input_utils.py
Comment (dfu-godaddy) [reply sample_size/scripts/input_utils.py:115]: Creating a standalone PR for this TODO makes sense to me. The priority of editing the scripts is low in terms of unblocking the rest of the calculator portal tickets.

File: sample_size/sample_size_calculator.py
Comment (dpandey-godaddy) [inline sample_size/sample_size_calculator.py:17]:
```python
with open("sample_size/metrics_schema.json", "r") as schema_file:
      METRICS_SCHEMA = json.load(open("sample_size/metrics_schema.json"))
```
good practice to use this

**Review by dpandey-godaddy — APPROVED** [review]
A'd after comment

File: sample_size/scripts/input_utils.py
Comment (pchiang-godaddy) [reply sample_size/scripts/input_utils.py:115]: it is probably better to release together? i don't think that TODO is hard to block the other work for so long though. just thinking we can then make the release once instead of making twice just due to that small TODO

Comment (xyzou685) [reply sample_size/scripts/input_utils.py:115]: Do you mean fixing it in this PR? I don't see a straightforward way to fix it since we cannot check whether the metric being registered is the first metric and customize questions based on that. If we simply want it to work properly can I just revert my previous change on this for now https://github.com/godaddy/sample-size/pull/30#discussion_r882216839, as it actually works, though not in the most efficient way...

Comment (pchiang-godaddy) [reply sample_size/scripts/input_utils.py:115]: Why is your original version not efficient? i don't see super big difference though

oh i just mean, either fix it in this PR, or create a new individual PR. Not with the PR which is to merge all your change with `main` branch so we know what is updated for this TODO. My opinion is a bit different with dannie, i guess she suggests you merging all the change and do the release without finishing this TODO to unblock other works. but I think this TODO is not really hard, we can just do it and do one release to support multiple metrics. instead of releasing a version with a broken (incomplete) script.

Comment (pchiang-godaddy) [reply sample_size/scripts/input_utils.py:115]: or you can do something similar to what Dileep has suggested:
e.g.
```python
def get_metrics() -> List[Dict[str, Any]]:
    metrics = []
    def get_metric():
        metric_type = get_metric_type()
        metric_metadata = get_metric_parameters(METRIC_PARAMETERS[metric_type])
        metric_metadata["mde"] = get_mde(metric_type)
        metrics.append({"metric_type": metric_type, "metric_metadata": metric_metadata})

    get_metric()
    while register_another_metric():
        get_metric()

    return metrics
```

File: tests/sample_size/test_sample_size_calculator.py
Comment (pchiang-godaddy) [inline tests/sample_size/test_sample_size_calculator.py:162]:
```suggestion
    # TODO: parameterize register metric functions
```

**Review by pchiang-godaddy — COMMENTED** [review]
with a nit.
For the script comment  https://github.com/godaddy/sample-size/pull/30#discussion_r884024670 - although I think it's probably easier to fix that in current PR, if people want to really make another PR I'm fine. It is just I don't think that is needed. And i suggest to do the release when this is fixed. So it doesn't matter if you fix here or make a new one. Unless people have any other opinions

**Review by pchiang-godaddy — APPROVED** [review]

**Pull request merged**
